civis/service_client.py

```diff
@@ -0,0 +1,140 @@
+from __future__ import absolute_import
+
+from collections import OrderedDict
+import re
+import warnings
+
+from jsonref import JsonRef
+import requests
+import six
+
+from civis import APIClient
+from civis.base import Endpoint, CivisAPIError, tostr_urljoin
+from civis.resources._resources import parse_method
+from civis._utils import to_camelcase
+
+
+def auth_service_session(session, service_id):
+    try:
+        client = APIClient()
+        service = client.services.get(service_id)
+    except CivisAPIError as err:
+        msg = ('There was an issue '
+               'finding service with ID {}.').format(service_id)
+        six.raise_from(ValueError(msg), err)
+
+    auth_url = service['current_deployment']['displayUrl']
+    # Make request for adding Authentication Cookie to session
+    session.get(auth_url)
+
+
+class ServiceEndpoint(Endpoint):
+
+    def __init__(self, session_kwargs, client,
+                 return_type='civis'):
+        self._session_kwargs = session_kwargs
+        self._return_type = return_type
+        self._client = client
+
+    def _build_path(self, path):
+        if not path:
+            return self._client._base_url
+        if not self._client._root_path:
+            return tostr_urljoin(self._client._base_url, path.strip("/"))
+        return tostr_urljoin(self._client._base_url,
+                             self._client._root_path.strip("/"),
+                             path.strip("/"))
+
+    def _make_request(self, method, path=None, params=None, data=None,
+                      **kwargs):
+        url = self._build_path(path)
+
+        with requests.Session() as sess:
+            auth_service_session(sess, self._client._service_id)
+            with self._lock:
+                response = sess.request(method, url, json=data,
+                                        params=params, **kwargs)
+
+        if not response.ok:
+            six.raise_from(ValueError(response.text),
+                           ValueError)
+
+        return response
+
+
+class ServiceClient():
+
+    def __init__(self, service_id, root_path=None, swagger_path="/endpoints"):
+        return_type = 'snake'
+        self._session_kwargs = {}
+        self._service_id = service_id
+        self._base_url = self.get_base_url()
+        self._root_path = root_path
+        self._swagger_path = swagger_path
+        # Catch deprecation warnings from generate_classes_maybe_cached and
+        # the functions it calls until the `resources` argument is removed.
```

Comment:
I don't understand this, can you say more?

Reply:
This was copied from APIClient(), generate_classes_maybe_cached() is a function with from depreciated parameter so this was here to catch warnings. But we are not using that function so I can take it out.
At the time I wasn't sure if I should keep generate_classes_maybe_cached() or not. It allows a user to pass in an api spec as a file or dict, instead of downloading one from an endpoint.
Do you have strong feelings about using it or not?

Reply:
I do not

Reply:
Seems like [YAGNI](https://martinfowler.com/bliki/Yagni.html) to me. I'd vote to take it out until we hear requests for it. I don't feel strongly on that though.

Reply:
I can comment on the motivation for allowing API client generation without having to hit a remote endpoint. It's been very useful for testing. Building an API client from a stored spec lets us test code that depends on the client without needing to hard-code structure (example: https://github.com/civisanalytics/civis-python/blob/v1.11.0/civis/tests/mocks.py#L30). It's also used for between-session caching in the CLI. Downloading the spec and regenerating endpoints can take a non-trivial amount of time. I don't know if those are relevant here. You could consider adding an `@lru_cache` decorator to `ServiceClient.generate_classes` to speed up ServiceClient instantiation within a session.

Reply:
Thanks for your Insight @stephen-hoover! I'll take a better look at our test cases and the decorator. It does take a bit of time to generate the client.

Reply:
Ah, that makes sense. In that case, it seems reasonable to keep (although we don't have anything to test it on quite yet).